File: packages/react-native/android/src/main/java/dev/openota/OtaConfig.kt
package dev.openota

import android.content.Context
import android.content.pm.PackageManager
import android.os.Bundle

/**
 * Values the config plugin (Expo) or the `ota init` codemod (bare RN) bakes into
 * the binary, as <meta-data> on <application>:
 *
 *   dev.openota.API_URL           https://ota.example.com
 *   dev.openota.APP_KEY           pk_...
 *   dev.openota.PROJECT_ID        prj_...
 *   dev.openota.CHANNEL           production
 *   dev.openota.RUNTIME_VERSION   fp_...            (@expo/fingerprint)
 *   dev.openota.PUBLIC_KEY        PEM or bare base64 SPKI
 *   dev.openota.EMBEDDED_FLOOR_ID UUIDv7 stamped at build time
 *   dev.openota.DEEP_LINK_SCHEME  myapp
 */
internal data class OtaConfig(
  val apiUrl: String,
  val appKey: String,
  val projectId: String,
  val channel: String,
  val runtimeVersion: String,
  val publicKey: String,
  val embeddedFloorId: String?,
  val nativeVersion: String,
  val deepLinkScheme: String?,
) {
  val isConfigured: Boolean get() = projectId.isNotEmpty() && publicKey.isNotEmpty()

  fun toMap(): Map<String, Any?> = mapOf(
    "apiUrl" to apiUrl,
    "appKey" to appKey,
    "projectId" to projectId,
    "channel" to channel,
    "runtimeVersion" to runtimeVersion,
    "publicKey" to publicKey,
    "embeddedFloorId" to embeddedFloorId,
    "nativeVersion" to nativeVersion,
    "deepLinkScheme" to deepLinkScheme,
  )

  companion object {
    private const val PREFIX = "dev.openota."

    fun from(context: Context): OtaConfig {
      val app = context.applicationContext
      val meta = runCatching {
        app.packageManager
          .getApplicationInfo(app.packageName, PackageManager.GET_META_DATA)
          .metaData
      }.getOrNull()

      val nativeVersion = runCatching {
        app.packageManager.getPackageInfo(app.packageName, 0).versionName
      }.getOrNull().orEmpty()

      return OtaConfig(
        apiUrl = meta.value(app, "API_URL").orEmpty().trimEnd('/'),
        appKey = meta.value(app, "APP_KEY").orEmpty(),
        projectId = meta.value(app, "PROJECT_ID").orEmpty(),
        channel = meta.value(app, "CHANNEL") ?: "production",
        runtimeVersion = meta.value(app, "RUNTIME_VERSION").orEmpty(),
        publicKey = meta.value(app, "PUBLIC_KEY").orEmpty(),
        embeddedFloorId = meta.value(app, "EMBEDDED_FLOOR_ID"),
        nativeVersion = nativeVersion,
        deepLinkScheme = meta.value(app, "DEEP_LINK_SCHEME"),
      )
    }

    @Suppress("DEPRECATION") // Bundle.get is the only untyped meta-data read
    private fun Bundle?.value(context: Context, key: String): String? =
      when (val raw = this?.get(PREFIX + key)) {
        is String -> raw.ifEmpty { null }
        // A plugin may emit android:value="@string/..." — meta-data then carries the resource id.
        is Int -> runCatching { context.getString(raw) }.getOrNull()
        null -> null
        else -> raw.toString()
      }
  }
}

File: packages/react-native/ios/OtaConfig.swift
import Foundation

/**
 Values the config plugin (Expo) or the `ota init` codemod (bare RN) bakes into
 the binary. iOS carries them as flat `OpenOta*` keys in Info.plist:

   <key>OpenOtaApiUrl</key>          <string>https://ota.example.com</string>
   <key>OpenOtaAppKey</key>          <string>pk_...</string>
   <key>OpenOtaProjectId</key>       <string>prj_...</string>
   <key>OpenOtaChannel</key>         <string>production</string>
   <key>OpenOtaRuntimeVersion</key>  <string>fp_...</string>   <!-- @expo/fingerprint -->
   <key>OpenOtaPublicKey</key>       <string>PEM or bare base64 SPKI</string>
   <key>OpenOtaEmbeddedFloorId</key> <string>UUIDv7 stamped at build time</string>
   <key>OpenOtaDeepLinkScheme</key>  <string>myapp</string>

 Flat on purpose: both writers — the Expo plugin's plist mod and the bare-RN
 XML codemod — insert simple key/string pairs, and a nested dict would give the
 XML codemod a structure to parse instead of a line to add. The Android side
 reads the same names as `dev.openota.<UPPER_SNAKE>` meta-data.
 Key names live in plugin/codemods/edits.js (IOS_PLIST) — keep in step.
 */
struct OtaConfig {
  let apiUrl: String
  let appKey: String
  let projectId: String
  let channel: String
  let runtimeVersion: String
  let publicKey: String
  let embeddedFloorId: String?
  let nativeVersion: String
  let deepLinkScheme: String?

  var isConfigured: Bool { !projectId.isEmpty && !publicKey.isEmpty }

  var asDictionary: [String: Any?] {
    [
      "apiUrl": apiUrl,
      "appKey": appKey,
      "projectId": projectId,
      "channel": channel,
      "runtimeVersion": runtimeVersion,
      "publicKey": publicKey,
      "embeddedFloorId": embeddedFloorId,
      "nativeVersion": nativeVersion,
      "deepLinkScheme": deepLinkScheme,
    ]
  }

  static func load(from bundle: Bundle = .main) -> OtaConfig {
    func value(_ key: String) -> String? {
      guard let text = bundle.object(forInfoDictionaryKey: "OpenOta" + key) as? String,
            !text.isEmpty
      else { return nil }
      return text
    }

    var apiUrl = value("ApiUrl") ?? ""
    while apiUrl.hasSuffix("/") { apiUrl.removeLast() }

    return OtaConfig(
      apiUrl: apiUrl,
      appKey: value("AppKey") ?? "",
      projectId: value("ProjectId") ?? "",
      channel: value("Channel") ?? "production",
      runtimeVersion: value("RuntimeVersion") ?? "",
      publicKey: value("PublicKey") ?? "",
      embeddedFloorId: value("EmbeddedFloorId"),
      nativeVersion: bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "",
      deepLinkScheme: value("DeepLinkScheme")
    )
  }
}
